toolchain/source/source_buffer.cpp

```diff
@@ -41,25 +83,21 @@ auto SourceBuffer::CreateFromFile(llvm::vfs::FileSystem& fs,
     emitter.Emit(filename, ErrorStattingFile, file.getError().message());
     return std::nullopt;
   }
-  int64_t size = status->getSize();
-  if (size >= std::numeric_limits<int32_t>::max()) {
-    CARBON_DIAGNOSTIC(FileTooLarge, Error,
-                      "File is over the 2GiB input limit; size is {0} bytes.",
-                      int64_t);
-    emitter.Emit(filename, FileTooLarge, size);
-    return std::nullopt;
-  }
+
+  // `stat` on a file without a known size gives a size of 0, which causes
+  // `llvm::vfs::File::getBuffer` to produce an empty buffer. Use a size of -1
+  // in this case so we get the complete file contents.
+  bool is_regular_file = status->isRegularFile();
+  int64_t size = is_regular_file ? status->getSize() : -1;
 
   llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> buffer =
       (*file)->getBuffer(filename, size, /*RequiresNullTerminator=*/false);
-  if (buffer.getError()) {
-    CARBON_DIAGNOSTIC(ErrorReadingFile, Error, "Error reading file: {0}",
-                      std::string);
-    emitter.Emit(filename, ErrorReadingFile, file.getError().message());
+
+  if (!CheckBuffer(buffer, "<stdin>", emitter)) {
```

Comment:
Pretty sure this should be:
```suggestion
  if (!CheckBuffer(buffer, filename, emitter)) {
```

Is it worth trying to craft a test of this? I don't see any good way sadly -- would need an "actual" file that was un-readable or over 2G which seems ... not friendly for our testdata. =]

Maybe we could test this in a unit-test? With somewhat minor platform-specific if-defs, it's not too hard to write a unittest that ends up failing either of these diagnostics... With the vfs layer it might be quite easy to get a virtual file with a huge buffer that is never materialized...

Reply:
I've added a test for the "can't open file" error. The easiest way I could find was opening a directory :)

I think we could test the other diagnostic via a unit test, if you think that's worth doing as part of this PR. We'd need to actually allocate 4GB of address space, which might hit a ulimit I suppose, but otherwise seems doable in a 64-bit-only test.

Reply:
I think this is pretty solid, mostly wanted to do whatever testing was easy enough given the virtual FS stuff we have. Thanks!